tests/test_build_nanoset_dataloader.py

```diff
@@ -0,0 +1,130 @@
+from nanotron.data.utils import compute_datasets_num_samples
+from nanotron.parallel import ParallelContext
+
+
+@pytest.mark.parametrize(
+    "tp,dp,pp",
+    [
+        pytest.param(*all_3d_configs)
+        for gpus in range(1, min(available_gpus(), 4) + 1)
+        for all_3d_configs in get_all_3d_configurations(gpus)
+    ],
+)
+@rerun_if_address_is_in_use()
+def test_build_nanoset_dataloader(tp: int, dp: int, pp: int):
+    test_context = TestContext()
+
+    # Create dataset files
+    json_paths, bin_paths = create_dataset_paths(tmp_dir=test_context.get_auto_remove_tmp_dir(), quantity=2)
+
+    # Create dummy json datasets
+    for idx, json_path in enumerate(json_paths):
+        create_dummy_json_dataset(path_to_json=json_path, dummy_text=f"Nanoset {idx}!")
+
+    # Preprocess dummy json datasets
+    for json_path in json_paths:
+        preprocess_dummy_dataset(path_to_json=json_path)
+
+    init_distributed(tp=tp, dp=dp, pp=pp)(_test_build_nanoset_dataloader)(
+        test_context=test_context, path_to_bin_files=bin_paths
+    )
+
+
+def _test_build_nanoset_dataloader(
+    parallel_context: ParallelContext, test_context: TestContext, path_to_bin_files: str
+):
+    SEED = 1234
+    random.seed(SEED)
+
+    TRAIN_STEPS = random.randint(1000, 10000)
+    VAL_CHECK_INTERVAL = TRAIN_STEPS / 4
+    VAL_STEPS = TRAIN_STEPS / 10
+    MICRO_BATCH_SIZE = 2 ** random.randint(1, 4)
+    N_MICRO_BATCHES_PER_BATCH = 2 ** random.randint(1, 4)
+    GLOBAL_BATCH_SIZE = MICRO_BATCH_SIZE * N_MICRO_BATCHES_PER_BATCH * parallel_context.dp_pg.size()
+
+    SEQ_LENGTH = 2 ** random.randint(10, 14)
+    SPLIT = "{},{},{}".format(random.randint(1, 100), random.randint(1, 100), random.randint(1, 100))
+
+    input_pp_rank, output_pp_rank = 0, int(parallel_context.pp_pg.size() - 1)
+
+    # Create Nanoset configs: 1. Normal 2. Blended
+    split_num_samples = compute_datasets_num_samples(
+        train_iters=TRAIN_STEPS,
+        eval_interval=VAL_CHECK_INTERVAL,
+        eval_iters=VAL_STEPS,
+        global_batch_size=GLOBAL_BATCH_SIZE,
+    )
+
+    nanoset_config = NanosetConfig(
+        random_seed=SEED,
+        sequence_length=SEQ_LENGTH,
+        data_path=path_to_bin_files[0],
+        split=SPLIT,
+        split_num_samples=split_num_samples,
+        path_to_cache=test_context.get_auto_remove_tmp_dir(),
+    )
+
+    blended_nanoset_config = NanosetConfig(
+        random_seed=SEED,
+        sequence_length=SEQ_LENGTH,
+        data_path={bin_path: random.randint(1, 100) for bin_path in path_to_bin_files},
+        split=SPLIT,
+        split_num_samples=split_num_samples,
+        path_to_cache=test_context.get_auto_remove_tmp_dir(),
+    )
+
+    configs = [nanoset_config, blended_nanoset_config]
+    dataset_types = [Nanoset, BlendedNanoset]
+
+    for config, dataset_type in zip(configs, dataset_types):
+        # Create Nanosets
+        train_dataset, valid_dataset, test_dataset = NanosetBuilder(config).build()
```

Comment:
We should add a separate test to check the dataset length, sequence length, and non-overlap (ensuring dataset[1] != dataset[2]) in a BlendedNanoSet.

Reply:
The `BlendedNanoset` essentially determines which samples to extract from each Nanoset based on the requested [number of samples](https://github.com/TJ-Solergibert/nanotron/blob/16f44d1b8177b38a7ae3a2322a912562dbdc244a/src/nanotron/data/blended_nanoset.py#L27) for the `BlendedNanoset`, the [weights](https://github.com/TJ-Solergibert/nanotron/blob/16f44d1b8177b38a7ae3a2322a912562dbdc244a/src/nanotron/data/blended_nanoset.py#L25) of each `Nanoset`, and their respective lengths in [`build_blending_indices`](https://github.com/TJ-Solergibert/nanotron/blob/16f44d1b8177b38a7ae3a2322a912562dbdc244a/src/nanotron/data/blended_nanoset.py#L129). As can be seen [here](https://github.com/TJ-Solergibert/nanotron/blob/16f44d1b8177b38a7ae3a2322a912562dbdc244a/src/nanotron/data/blended_nanoset.py#L199), when establishing the index of the sample to extract from a specific `Nanoset`, the modulo operation with the length of the `Nanoset` is applied, so we NEVER attempt to access an index greater than the length of the `Nanoset`.

The number of samples required for the `BlendedNanoset` is computed during the initialization of [`NanosetConfig`](https://github.com/TJ-Solergibert/nanotron/blob/16f44d1b8177b38a7ae3a2322a912562dbdc244a/src/nanotron/data/nanoset_configs.py#L54), and the number of samples each` Nanoset` in the blend must have is computed [here](https://github.com/TJ-Solergibert/nanotron/blob/16f44d1b8177b38a7ae3a2322a912562dbdc244a/src/nanotron/data/dataset_builder.py#L58). Note that you can create a `Nanoset` from a file containing 10 documents, resulting in only 5 samples, and request it to have 200; in this case, it will only increase the number of times the indices are [concatenated](https://github.com/TJ-Solergibert/nanotron/blob/16f44d1b8177b38a7ae3a2322a912562dbdc244a/src/nanotron/data/nanoset.py#L236) (You will perform 40 epochs).

Reply:
> The BlendedNanoset essentially determines which samples to extract from each Nanoset based on the requested number of samples for the BlendedNanoset
Oops. Thanks for clarifying.. We should add the suggested test above for `Nanoset`